File: 39-combination-sum/39-combination-sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        res, N = [], len(candidates)
        
        def btrack(i, tot, cur):
            if i >= N or tot > target:
                return 0
            if tot == target:
                res.append(cur)
                return
            for j in range(i, N):
                x = candidates[j]
                btrack(j, tot+x, cur+[x])
            return res
        
        return btrack(0, 0, [])

File: 39-combination-sum/39-combination-sum.cpp
class Solution {
public:
    vector<vector<int>> res;
    vector<int> cur;
    set<set<int>> vis;
    
    void dfs(int idx, int sum, int target, vector<int>& cand) {
        if (sum >= target || idx >= cand.size()) {
            set<int> st(cur.begin(),cur.end());
            if (sum == target && !vis.count(st))
                res.push_back(cur);
            return;
        }
        for (int i = idx; i < cand.size(); i++) {
            cur.push_back(cand[i]);
            dfs(i, sum + cand[i], target, cand);
            cur.pop_back();
        }
    }
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        dfs(0, 0, target, candidates);
        return res;
    }
};